user/controller.py

```diff
@@ -3,76 +3,89 @@
-    if isinstance(users, list):
-        users_json = [json.loads(user.to_json()) for user in users]
-    else:
-        users_json = json.loads(users.to_json())
+        get_user_result = user_service.get(user_id)
+        # HACK: this serializes, deserializes, and then reserializes user. we really need dedicated serialization logic
+        # TODO: currently there are no access controls so this leaks password hashes/salts. also we are at the mercy of mongoengine's serialization logic which does weird things
+        if isinstance(get_user_result, list):
+            users_json = [json.loads(user.to_json()) for user in get_user_result]
+        else:
+            users_json = json.loads(get_user_result.to_json())
+        return jsonify({"users": users_json}), 200
+    except ValidationException as e:
+        return jsonify({"Validation error": str(e)}), 400
+    except Exception as e:
+        return jsonify({"Unknown error" : str(e)}), 500
 
-    return jsonify({"users": users_json}), 200
 
 @bp.route('/create', methods=['POST'])
 def create():
-    username = request.json['username']
-    password = request.json['password']
-
     try:
-        user_id = user_service.create(username, password)
-        return jsonify({"user.id": user_id}), 200
-    except DBException as e:
-        return jsonify({"error": e.message}), 400
+        username = str(request.json['username'])
+        password = str(request.json['password'])
+    except Exception as e:
+        return jsonify({"Request error" : f"{e}: {str(e)}"}), 400
+    try:
+        return jsonify({"user.id": user_service.create(username, password)}), 200
+    except ValidationException as e:
+        return jsonify({"Validation error": str(e)}), 400
+    except Exception as e:
+        return jsonify({"Unknown error" : str(e)}), 500
 
 @bp.route('/update', methods=['PATCH'])
 def update():
-    user_id = request.json['id']
-    username = request.json['username']
-    password = request.json['password']
-
+    try:
+        user_id = str(request.json['id'])
+        username = str(request.json['username'])
+        password = str(request.json['password'])
+    except Exception as e:
+        return jsonify({"Request error" : f"{e}: {str(e)}"}), 400
     try:
         user_service.update(user_id, username, password)
         return "Success", 200
-    except DBException as e:
-        return jsonify({"error": e.message}), 400
+    except ValidationException as e:
+        return jsonify({"Validation error": str(e)}), 400
+    except Exception as e:
+        return jsonify({"Unknown error" : str(e)}), 500
 
 @bp.route('/delete', methods=['DELETE'])
 def delete():
-    user_id = request.json['id']
-
+    try:
+        user_id = str(request.json['id'])
+    except Exception as e:
+        return jsonify({"Request error" : f"{e}: {str(e)}"}), 400
```

Comment:
for the record, `f"{e}" results in exactly the same thing as `f"{str(e)}"` so this will just result in the same thing twice delimited by a colon

Reply:
Just to set the record straight I was actually wrong here, `f"{e}"` should behave the same as `repr(e)` rather than `str(e)` but the two are similar enough in this case that it's still pretty redundant.